iotfunctions/anomaly.py

```diff
@@ -507,15 +509,36 @@ def prepare_data(self, dfEntity):
     def execute(self, df):
 
         df_copy = df.copy()
-        entities = np.unique(df_copy.index.levels[0])
-        logger.debug(str(entities))
 
-        df_copy[self.output_item] = 0
-        # df_copy.sort_index() - NoOP
+        logger.info('Expression exp: ' + self.expression + '  input: ' + str(self.input_item))
+        expr = self.expression
+        if '${' in expr:
+            expr = re.sub(r"\$\{(\w+)\}", r"df['\1']", expr)
+            msg = 'Expression converted to %s. ' % expr
+        else:
+            msg = 'Expression (%s). ' % expr
+
+        self.trace_append(msg)
+
+        expr = str(expr)
+        logger.info('Expression - after regexp: ' + expr)
+
+        try:
+            mask = eval(expr)
+            entities = df_copy.loc[mask].index.unique(level=0)
```

Comment:
FYI, if mask is a boolean series of same length as df_copy then the .loc() is not required. Actually, df_copy.loc[mask] is identical to df_copy[mask]. 
df_copy[] is threefold overloaded, accepting a boolean series or a column name or an array of column names